posthog/hogql/printer.py

```diff
@@ -278,12 +271,11 @@ def visit_join_expr(self, node: ast.JoinExpr) -> JoinExprResponse:
             if self.dialect == "clickhouse":
                 sql = table_type.table.to_printed_clickhouse(self.context)
 
-                # Always put S3 Tables in a CTE so joins can work when queried
-                if isinstance(table_type.table, S3Table):
-                    cte = f"{self._print_identifier(node.alias)} AS (SELECT * FROM {sql})"
-
-                    # The table is captured in a CTE so just print the table name in the final select
-                    sql = self._print_identifier(node.alias or table_type.table.name)
+                # Edge case. If we are joining an s3 table, we must wrap it in a subquery for the join to work
+                if isinstance(table_type.table, S3Table) and (
+                    node.next_join or node.join_type == "JOIN" or node.join_type == "GLOBAL JOIN"
+                ):
+                    sql = f"(SELECT * FROM {sql})"
```

Comment:
This _might_ be slower because of the asterisk resulting in all columns being read. Ideally we'd only select the relevant columns here. But not sure how ClickHouse reads tables from S3, so can't say if there's an impact with certainty

Reply:
I'm keeping this in mind as I think you're right but I don't think it's worth building the pushdown logic here yet as most external datasets will probably be small atm. Also, most of this data modeling will be for internal use and a controlled MVP rollout

Reply:
Yup, that's fine